refactor: introduce expected pattern for error handling in master service

xiaguan opened the pull request:
## Overview

This PR introduces the `tl::expected` pattern for error handling in the master service, replacing traditional `ErrorCode` return types with a more type-safe and explicit approach.

## Changes

### Core Improvements
- **Type-safe error handling**: Replace `ErrorCode` return types with `tl::expected<T, ErrorCode>` pattern
- **Clearer API semantics**: Success values and error codes are now explicitly separated
- **Better developer experience**: 
  - Success cases: access via `.value()`
  - Error cases: access via `.error()`
  - Eliminates ambiguity between success and error states

### Modified Components
- **MasterService methods**: Updated to return `expected<void, ErrorCode>` or `expected<T, ErrorCode>`
- **RPC service interfaces**: Modified to support the expected pattern
- **Test suites**: Updated all related tests to handle new expected return types
- **Dependencies**: Added necessary includes for `ylt/util/expected.hpp`

### Files Changed
- `mooncake-store/include/master_service.h` - Updated method signatures
- `mooncake-store/include/rpc_service.h` - RPC interface adaptations
- `mooncake-store/include/types.h` - Minor formatting improvements
- `mooncake-store/src/master_service.cpp` - Implementation updates
- `mooncake-store/tests/` - Comprehensive test updates

## Benefits

1. **Type Safety**: Compile-time guarantees about error handling
2. **Explicit Error Handling**: No more ambiguous return codes
3. **Better Maintainability**: Clear separation of success and error paths
4. **Future-Ready**: Foundation for extending expected pattern to RPC responses

## Future Work

This PR lays the groundwork for:
- Extending expected pattern to RPC response types
- Enhancing the error code system for more comprehensive error handling
- Further improving type safety across the codebase

## Testing

All existing tests have been updated and continue to pass, ensuring backward compatibility and correctness of the new error handling approach.

---

**Note**: This change maintains API compatibility while improving error handling semantics. The expected pattern provides a modern, type-safe approach to error handling that will benefit future development and maintenance.

ykwd commented on mooncake-store/include/rpc_helper.h:
```diff
@@ -0,0 +1,59 @@
+        // Special handling for bool to log the value, similar to the original
+        // implementation of ExistKey.
+        if constexpr (std::is_same_v<ResultType, bool>) {
+            timer.LogResponse("success=true, exist=", result.value());
```
1. Using success=true or false is a little bit ambiguous. Also, the "exist=" may only be suitable if the returned value denotes whether something exists or not. How about timer.LogResponse("status=success, return=", result.value());

ykwd commented on mooncake-store/include/rpc_helper.h:
```diff
@@ -0,0 +1,59 @@
+        if constexpr (std::is_same_v<ResultType, bool>) {
+            timer.LogResponse("success=true, exist=", result.value());
+        } else {
+            timer.LogResponse("success=true");
```
timer.LogResponse("status=success“);

ykwd commented on mooncake-store/include/rpc_helper.h:
```diff
@@ -0,0 +1,59 @@
+    } else {
+        inc_fail_metric();
+        timer.LogResponse("success=false, error=", toString(result.error()));
+    }
```
timer.LogResponse("status=fail, error=", toString(result.error()));

xiaguan replied:
resolved

ykwd commented on mooncake-store/include/rpc_helper.h:
```diff
@@ -0,0 +1,59 @@
+auto execute_rpc(std::string_view rpc_name, RpcCallable&& rpc_call,
+                 LogRequestCallable&& log_request,
+                 std::function<void()> inc_req_metric,
+                 std::function<void()> inc_fail_metric) {
```
Just like the rpc_call and log_request, why don't we also use template for in_req_metric and inc_fail_metric? The template can prevent the overhead of the std::function

ykwd commented on mooncake-store/include/rpc_service.h:
```diff
@@ -260,310 +198,231 @@ class WrappedMasterService {
-    }
+        std::vector<tl::expected<std::vector<Replica::Descriptor>, ErrorCode>>
+            results;
+        results.reserve(keys.size());
```
Will we record batchGet and batchPut metrics?

ykwd reviewed (commented):
Great work! Just a few suggestions in the comments.

xiaguan committed:
fix test compile

xiaguan committed:
refactor: introduce expected pattern for error handling in master service

- Replace ErrorCode return types with tl::expected<T, ErrorCode> pattern
- Improve error handling clarity by separating success values from error codes
- Update MasterService methods to return expected<void, ErrorCode> or expected<T, ErrorCode>
- Modify RPC service interfaces to support expected pattern
- Update all related tests to handle new expected return types
- Add necessary includes for ylt/util/expected.hpp

This change makes error handling more explicit and type-safe:
- Success cases can be accessed via .value()
- Error cases can be accessed via .error()
- Eliminates ambiguity between success and error states

Future work:
- Extend expected pattern to RPC response types
- Enhance error code system for more comprehensive error handling

xiaguan committed:
fix: update Python integration to work with new batch API

- Replace BatchObjectInfo with vector<vector<Replica::Descriptor>>
- Update BatchPut to handle new return type vector<tl::expected<void, ErrorCode>>
- Fix BatchQuery API usage to work with new expected pattern
- Convert unordered_map to vector format for BatchPut parameter compatibility

xiaguan committed:
refactor: update client implementation and remove master.proto

- Enhanced client.h and client.cpp with new functionality
- Removed obsolete master.proto file
- Updated master_client.cpp and transfer_task.cpp
- Improved integration and stress tests

xiaguan committed:
refine master log and metric

xiaguan committed:
Refactor error handling to use expected<T,E> pattern instead of ErrorCode

- Updated MasterService methods to replace ylt::expected with tl::expected for better error handling.
- Modified BatchGetReplicaList, BatchPutStart, BatchPutEnd, and other methods to return tl::expected types.
- Enhanced ClientIntegrationTest to handle expected results from Put, Get, Remove, and other operations using tl::expected.
- Adjusted error handling in clientctl and master_metrics_test to utilize the new expected type.
- Improved overall error reporting in tests to provide clearer feedback on operation failures.

xiaguan committed:
fix(ci): should alloc first

xiaguan committed:
Merge branch 'main' into master_expected

Resolved merge conflicts in multiple files:
- mooncake-integration/store/store_py.cpp: Updated replica handling to support both memory and disk replicas
- mooncake-store/include/client.h: Unified API signatures for Get and BatchPut methods
- mooncake-store/include/rpc_service.h: Added response structures and fixed Ping method
- mooncake-store/src/client.cpp: Adopted main branch implementation with storage backend support
- mooncake-store/src/master_client.cpp: Used main branch version
- mooncake-store/src/transfer_task.cpp: Used main branch version
- mooncake-store/tests/*: Used main branch test implementations

Key changes integrated:
- Enhanced storage backend functionality
- Ascend transport support
- Improved error handling and batch operations
- Updated test infrastructure

xiaguan committed:
merge main

xiaguan committed:
fix tests
